tfjs-tflite/README.md

```diff
@@ -0,0 +1,123 @@
+# TFLite support for Tensorflow.js
+
+_WORK IN PROGRESS_
+
+This package enables users to run arbitary TFLite models on the web. Users can
+load a TFLite model from a URL, use TFJS tensors to set the model's input
+data, run inference, and get the output back in TFJS tensors. Under the hood,
+the TFLite C++ runtime is packaged in a set of WASM modules, and the one with
+the best performance will be automatically loaded based on user's current
+environment (e.g. whether WebAssembly SIMD and/or multi-threading is supported
+or not).
+
+Check out this [demo][demo] where we use this package to run a
+[CartoonGAN][model] TFLite model on the web.
+
+# Usage
+
+## Import the packages
+
+To use this package, you will need a TFJS backend installed. We recommend the
+CPU backend. You will also need to import `@tensorflow/tfjs-core` for
+manipulating tensors.
+
+### Via NPM
+
+```js
+// Adds the CPU backend.
+import '@tensorflow/tfjs-backend-cpu';
+// Import @tensorflow/tfjs-core
+import * as tf from '@tensorflow/tfjs-core';
+// Import @tensorflow/tfjs-tflite.
+import {loadTFLiteModel, TFLiteModel} from '@tensorflow/tfjs-tflite';
+```
+
+### Via a script tag
+
+```html
+<!-- Adds the CPU backend -->
+<script src="https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-backend-cpu"></script>
+<!-- Import @tensorflow/tfjs-core -->
+<script src="https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-core"></script>
+<!-- Import @tensorflow/tfjs-tflite -->
+<script src="https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-tflite"></script>
+```
+
+## Set WASM modules location (optional)
+
+By default, it will try to load the WASM modules from the same location where
+the package or your own script is served. Use `setWasmPath` to set your own
+location. See `src/tfweb_client.d.ts` for more details.
+
+
+```js
+import {setWasmPath} from '@tensorflow/tfjs-tflite';
+
+setWasmPath('https://your-server/path');
+```
+
+## Load a TFLite model
+```js
+const tfliteModel = await loadTFLiteModel('url/to/your/model.tflite');
+```
+
+## Run inference
+```js
+// Prepare input tensors.
+const img = tf.browser.fromPixels(document.querySelector('img'));
+const input = tf.sub(tf.div(tf.expandDims(img), 127.5), 1);
+
+// Run inference and get output tensors.
+let outputTensor = tfliteModel.predict(input) as tf.Tensor;
+console.log(outputTensor.dataSync());
+```
+
+# Performance
+
+Similar to TFJS WASM backend, this package uses [XNNPACK][xnnpack] to accelerate
+model inference. To achieve the best performance, use a browser that supports
+"WebAssembly SIMD" and "WebAssembly threads". In Chrome, these can be enabled
+in `chrome://flags/`. As of March 2021, XNNPACK can only be enabled for
+non-quantized TFLite models. Quantized models can still be used, but not
+accelerated. Support for quantized model acceleration is in the works.
```

Comment:
@jinjingforever is there work being done actively on supporting quantized model acceleration with XNNPACK? This sentence makes that seem like a matter of fact, but I haven't seen any open PRs or documentation about this in the various repos I've looked. Where does the work need to be done?

Our model is optimized in a post-training step to quantize our float parameters down to Int8s – we were able to see a significant speedup on Android but not on the web yet, unsurprisingly.

Our ideal workflow would actually just be to build tflite, XNNPACK, and an interface for our model for wasm directly with emscripten, but that doesn't seem to be documented yet either.

Reply:
Thanks for bringing this up Geordie! I asked the main engineer of xnnpack about this, and here is the info I got:

Currently, xnnpack already supports quantized tflite models (only in WASM SIMD, not in WASM MVP) with the following limitations:

1. Support for quantized inference is off by default. We need to enable it with a flag. This is something the tfjs team will take a look and work on. Will enable this before we officially publish this package.
2. Only signed per-tensor quantization schema is supported. This is what you get for models using quantization-aware training in TF MOT.
3. Only 4 operators are supported: ADD, CONV_2D, DEPTHWISE_CONV_2D, FULLY_CONNECTED. There is ongoing effort to support more operators.

Thanks!

Reply:
Thanks a lot for the info @jinjingforever, that's really helpful!

I'm not 100% sure but I can imagine that the results of the [post-training dynamic range quantization](https://www.tensorflow.org/lite/performance/post_training_quant) we are using also follows the signed per-tensor quantization schema (which I assume means [this](https://www.tensorflow.org/lite/performance/quantization_spec)). Most of the work our model is doing will be via `CONV_2D`, so that sounds promising.

It would be really awesome to try this out! It wasn't clear to me in this PR how we are building tflite for wasm. This PR appears to download pre-built binaries from gcp, which makes a lot of sense for more general use. Would it be possible to get some documentation of how those binaries are built and which flag would be necessary to enable the optimizations? Even a few scraps of knowledge would go a long way :) I did make some headway hacking a wasm config into tensorflow's Bazel `BUILD`, but since there is now an "official" build it seems that a cleaner way must exist.

I would also understand if this process is already in full swing and I just need to wait a while. I am totally aware that good work takes time and patience. Just want to express clear interest in the project :)

Reply:
Thanks for your understanding (and interests in this project!) :) The wasm binary is built internally like you said. We do plan to open-source the code in the near future, but it might take some time. For now, we will try to enable that flag when we built the wasm binary internally and do some experiments ourselves as well. We hope to make this part work before IO. 

Thanks!

Reply:
Awesome, best of luck to you all! I look forward to the result.

Many thanks

Reply:
Hi Geordie,

Sorry for the delay.. I finally got some time and I have enabled the flag I mentioned previously. You can give it a try using the [0.0.1-alpha.4](https://www.npmjs.com/package/@tensorflow/tfjs-tflite) package. 

Unfortunately, I wasn't able to test it because there is no publicly available models that can take advantage of the optimization. I talked to the engineer and it seems that this feature is mostly used internally for now. But please feel free to try it on your models. 

We will continue to work with the XNNPACK team to make this more usable externally. Will definitely keep the community updated. Thank you!

Reply:
@jinjingforever thanks a lot! I was away for a couple of weeks so sorry for my late reply as well. We are excited to try out the XNNPACK optimisations for wasm, hopefully we will get a chance in the next week or so.

Ultimately it would be ideal for our use case if we could build tflite and XNNPACK for emscripten directly and skip tfjs entirely. The goal would be a "black box" wasm function (including our custom native code) that takes the raw input and returns the model output. That way our feature extraction and the model prediction could all run "natively" without any bridging via javascript – that should provide the best performance and lowest overhead.

We are comfortable doing some level of building things ourselves to achieve that, but at the moment it seems like there is no publicly supported way to build tflite and XNNPACK for web yet. Again, we totally understand the tf teams have their own goals and deadlines – just wanted to express interest in that use case (and I doubt we'd be alone there) with the hope of having some "publicly-supported wasm build" in the longer term.

Thank you again! You're doing great work